msprime/trees.py

```diff
@@ -159,6 +177,34 @@ def __str__(self):
 
 
 class EdgesetTable(_msprime.EdgesetTable):
+    """
+    Class for tables describing all edgesets in a tree sequence, of the form
+        left	right	parent	children
+        5.0	7.0	4	0,1
+        2.0	2.5	5	1,3
+        3.2	3.5	4	0,1
+        0.0	3.2	6	0,1,3
+    These describe the half-open genomic interval affected: `[left, right)`,
+    the `parent` and the `children` on that interval.
+
+    Requirements: to describe a valid tree sequence, a `EdgesetTable` (and
+    corresponding `NodeTable`, to provide birth times) must satisfy:
+        1. each list of children must be in sorted order,
+        2. any two edgesets that share a child must be nonoverlapping, and
+        3. the birth times of the `parent` in an edgeset must be strictly
+            greater than the birth times of the `children` in that edgeset.
+    Furthermore, for algorithmic requirements
+        4. the the table must be sorted by birth time of the `parent`, and
+        5. any two edgesets corresponding to the same `parent` must be nonoverlapping.
+    It is an additional requirement that the complete ancestry of each sample
+    must be specified, but this is harder to verify.
+
+    It is not required that all records corresponding to the same parent be
```

Comment:
I ran some quick non-exhaustive checks and did not come across any errors, so I think it is not currently a requirement.

Hm.  I agree, this would be useful!  And, "gathering all edges for a given parent" is something that is not yet efficiently implemented, right?

I was thinking that not having this requirement would make the forwards-time algorithm easier, since we could just push every new individual's record onto the output stack, but in fact I was wrong: with overlapping generations we'll still need to sort on parent birth time.

I am already advocating for having two sets of requirements, "strict" and "relaxed", with "relaxed" tree sequences not needing to satisfy the last requirement, but with `simplify` and `simulate` guaranteed to output strict tree sequences, but methods other than `simplify` not guaranteed to work on relaxed ones.  So one option would be to add more requirements to `strict` ones - I think that edgesets are already left-to-right ordered in output, no?

So, how about this: there will need to be a method that orders edgesets appropriately: by parent birth time at least; possibly by left endpoint as well.  The only question is whether that is the user's responsibility or is the first step in simplify().  Either is fine, although it seems nicer and cleaner to put it in simplify, as otherwise we'll have to check if its true anyhow?  What do you think?